base/applications/mspaint/dib.cpp

```diff
@@ -142,148 +142,146 @@ GetDIBHeight(HBITMAP hBitmap)
+BOOL SaveDIBToFile(HBITMAP hBitmap, LPCWSTR FileName, BOOL fIsMainFile)
 {
     CImageDx img;
     img.Attach(hBitmap);
-    img.SaveDx(FileName, GUID_NULL, g_xDpi, g_yDpi); // TODO: error handling
+    HRESULT hr = img.SaveDx(FileName, GUID_NULL, g_xDpi, g_yDpi);
     img.Detach();
 
-    WIN32_FIND_DATA find;
-    HANDLE hFind = FindFirstFile(FileName, &find);
-    if (hFind == INVALID_HANDLE_VALUE)
+    if (FAILED(hr))
     {
-        ShowFileLoadError(FileName);
+        ShowError(IDS_SAVEERROR, FileName);
         return FALSE;
     }
-    FindClose(hFind);
 
-    // update time and size
-    FILETIME ft;
-    FileTimeToLocalFileTime(&find.ftLastWriteTime, &ft);
-    FileTimeToSystemTime(&ft, &g_fileTime);
-    g_fileSize = find.nFileSizeLow;
+    if (!fIsMainFile)
+        return TRUE;
 
-    // TODO: update hRes and vRes
-
-    registrySettings.SetMostRecentFile(FileName);
+    WIN32_FIND_DATAW find;
+    HANDLE hFind = ::FindFirstFileW(FileName, &find);
+    if (hFind == INVALID_HANDLE_VALUE)
+    {
+        ShowError(IDS_SAVEERROR, FileName);
+        return FALSE;
+    }
+    ::FindClose(hFind);
 
-    g_isAFile = TRUE;
+    SetFileInfo(FileName, &find, TRUE);
     g_imageSaved = TRUE;
     return TRUE;
 }
 
-void ShowFileLoadError(LPCTSTR name)
+void SetFileInfo(LPCWSTR name, LPWIN32_FIND_DATAW pFound, BOOL isAFile)
 {
-    CString strText;
-    strText.Format(IDS_LOADERRORTEXT, (LPCTSTR) name);
-    CString strProgramName;
-    strProgramName.LoadString(IDS_PROGRAMNAME);
-    mainWindow.MessageBox(strText, strProgramName, MB_OK | MB_ICONEXCLAMATION);
-}
-
-HBITMAP SetBitmapAndInfo(HBITMAP hBitmap, LPCTSTR name, DWORD dwFileSize, BOOL isFile)
-{
-    if (hBitmap == NULL)
+    // update file time and size
+    if (pFound)
     {
-        COLORREF white = RGB(255, 255, 255);
-        hBitmap = CreateColorDIB(registrySettings.BMPWidth,
-                                 registrySettings.BMPHeight, white);
-        if (hBitmap == NULL)
-            return FALSE;
-
-        HDC hScreenDC = GetDC(NULL);
-        g_xDpi = GetDeviceCaps(hScreenDC, LOGPIXELSX);
-        g_yDpi = GetDeviceCaps(hScreenDC, LOGPIXELSY);
-        ReleaseDC(NULL, hScreenDC);
+        FILETIME ft;
+        ::FileTimeToLocalFileTime(&pFound->ftLastWriteTime, &ft);
+        ::FileTimeToSystemTime(&ft, &g_fileTime);
 
+        g_fileSize = pFound->nFileSizeLow;
+    }
+    else
+    {
         ZeroMemory(&g_fileTime, sizeof(g_fileTime));
+        g_fileSize = 0;
     }
 
-    // update image
-    imageModel.PushImageForUndo(hBitmap);
-    imageModel.ClearHistory();
-
-    // update g_fileSize
-    g_fileSize = dwFileSize;
-
     // update g_szFileName
     if (name && name[0])
-        GetFullPathName(name, _countof(g_szFileName), g_szFileName, NULL);
+    {
+        CStringW strName = name;
+        ::GetFullPathNameW(strName, _countof(g_szFileName), g_szFileName, NULL);
+        // The following code won't work correctly when (name == g_szFileName):
```

Comment:
Again, such comment is not really useful, as it doesn't explain the actual reason why.

Reply:
Due to being the same pointer, storing data will corrupt.